File: engine/runtime-compiler/src/main/scala/org/enso/compiler/pass/analyse/alias/AliasMetadata.scala
package org.enso.compiler.pass.analyse.alias

import org.enso.compiler.pass.IRPass
import org.enso.compiler.pass.analyse.alias.graph.Graph
import org.enso.compiler.pass.analyse.alias.graph.GraphBuilder

/** Information about the aliasing state for a given IR node. */
sealed trait AliasMetadata extends IRPass.IRMetadata {

  /** The aliasing graph. */
  val graph: Graph
}

object AliasMetadata {
  sealed trait Scope extends AliasMetadata

  /** Aliasing information for a root scope.
    *
    * A root scope has a 1:1 correspondence with a top-level binding.
    *
    * @param graph the graph containing the alias information for that node
    */
  sealed case class RootScope(override val graph: Graph) extends Scope {
    override val metadataName: String = "AliasMetadata.RootScope"

    /** @inheritdoc */
    override def prepareForSerialization(
      compiler: Compiler
    ): RootScope = this

    /** @inheritdoc */
    override def restoreFromSerialization(
      compiler: Compiler
    ): Option[RootScope] = Some(this)

    /** @inheritdoc */
    override def duplicate(): Option[IRPass.IRMetadata] = None
  }

  /** Aliasing information about a child scope.
    *
    * @param graph the graph
    * @param scope the child scope in `graph`
    */
  sealed case class ChildScope(
    override val graph: Graph,
    scope: Graph.Scope
  ) extends Scope {
    def this(b: GraphBuilder) = {
      this(b.toGraph(), b.toScope())
    }

    override val metadataName: String = "AliasMetadata.ChildScope"

    /** @inheritdoc */
    override def prepareForSerialization(
      compiler: Compiler
    ): ChildScope = this

    /** @inheritdoc */
    override def restoreFromSerialization(
      compiler: Compiler
    ): Option[ChildScope] = Some(this)

    /** @inheritdoc */
    override def duplicate(): Option[IRPass.IRMetadata] = None
  }

  /** Aliasing information for a piece of [[IR]] that is contained within a
    * [[Scope]].
    *
    * @param graph the graph in which this IR node can be found
    * @param id the identifier of this IR node in `graph`
    */
  sealed case class Occurrence(
    override val graph: Graph,
    id: Graph.Id
  ) extends AliasMetadata {
    override val metadataName: String = "AliasMetadata.Occurrence"

    /** @inheritdoc */
    override def prepareForSerialization(
      compiler: Compiler
    ): Occurrence = this

    /** @inheritdoc */
    override def restoreFromSerialization(
      compiler: Compiler
    ): Option[Occurrence] = Some(this)

    /** @inheritdoc */
    override def duplicate(): Option[IRPass.IRMetadata] = None
  }
}

File: engine/runtime-compiler/src/main/java/org/enso/compiler/pass/analyse/alias/AliasMetadata.java
package org.enso.compiler.pass.analyse.alias;

import org.enso.compiler.core.CompilerStub;
import org.enso.compiler.core.IR;
import org.enso.compiler.core.ir.ProcessingPass;
import org.enso.compiler.pass.IRPass;
import org.enso.compiler.pass.analyse.AliasAnalysis;
import org.enso.compiler.pass.analyse.AliasAnalysis$;
import org.enso.compiler.pass.analyse.alias.graph.Graph;
import org.enso.compiler.pass.analyse.alias.graph.GraphBuilder;
import scala.Option;

/** Metata for {@link AliasAnalysis} pass. */
@SuppressWarnings("unchecked")
public abstract sealed class AliasMetadata implements IRPass.IRMetadata
    permits AliasMetadata.Scope, AliasMetadata.Occurrence {
  private final Graph graph;

  private AliasMetadata(Graph g) {
    this.graph = g;
  }

  public static <I extends IR> I updateMetadata(I ir, AliasMetadata data) {
    ir.passData().update(AliasAnalysis$.MODULE$, data);
    return ir;
  }

  public final Graph graph() {
    return graph;
  }

  @Override
  public final ProcessingPass.Metadata prepareForSerialization(CompilerStub compiler) {
    return this;
  }

  @Override
  public final Option<ProcessingPass.Metadata> restoreFromSerialization(CompilerStub compiler) {
    return Option.apply(this);
  }

  @Override
  public final Option duplicate() {
    return Option.empty();
  }

  public abstract static sealed class Scope extends AliasMetadata permits RootScope, ChildScope {
    private Scope(Graph g) {
      super(g);
    }
  }

  /**
   * Aliasing information for a root scope.
   *
   * <p>A root scope has a 1:1 correspondence with a top-level binding.
   *
   * @param graph the graph containing the alias information for that node
   */
  public static final class RootScope extends Scope {
    public RootScope(Graph graph) {
      super(graph);
    }

    @Override
    public String metadataName() {
      return "AliasMetadata.RootScope";
    }

    public final RootScope copy(Graph graph) {
      return new RootScope(graph);
    }
  }

  /** Aliasing information about a child scope. */
  public static final class ChildScope extends Scope {
    private final Graph.Scope scope;

    /**
     * Aliasing information about a child scope.
     *
     * @param b the graph builder
     */
    public static ChildScope from(GraphBuilder b) {
      return new ChildScope(b.toGraph(), b.toScope());
    }

    /**
     * Aliasing information about a child scope.
     *
     * @param graph the graph
     * @param scope the child scope in `graph`
     */
    public ChildScope(Graph graph, Graph.Scope scope) {
      super(graph);
      this.scope = scope;
    }

    public final Graph.Scope scope() {
      return scope;
    }

    public final ChildScope copy(Graph graph, Graph.Scope scope) {
      return new ChildScope(graph, scope);
    }

    @Override
    public String metadataName() {
      return "AliasMetadata.ChildScope";
    }
  }

  /** Aliasing information for a piece of [[IR]] that is contained within a [[Scope]]. */
  public static final class Occurrence extends AliasMetadata {
    private final int id;

    /**
     * Aliasing information for a piece of [[IR]] that is contained within a [[Scope]].
     *
     * @param graph the graph in which this IR node can be found
     * @param id the identifier of this IR node in `graph`
     */
    public Occurrence(Graph graph, int id) {
      super(graph);
      this.id = id;
    }

    @Override
    public String metadataName() {
      return "AliasMetadata.Occurrence";
    }

    public final Occurrence copy(Graph graph, int id) {
      return new Occurrence(graph, id);
    }

    public final Occurrence copy(Graph graph) {
      return new Occurrence(graph, this.id);
    }

    public int id() {
      return id;
    }
  }
}
